libyul/backends/evm/ControlFlowGraph.h

```diff
@@ -0,0 +1,193 @@
+}
+
+/// Control flow graph consisting of ``DFG::BasicBlock``s connected by control flow.
+struct CFG
+{
+	explicit CFG() {}
+	CFG(CFG const&) = delete;
+	CFG(CFG&&) = delete;
+	CFG& operator=(CFG const&) = delete;
+	CFG& operator=(CFG&&) = delete;
+
+	struct BuiltinCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<BuiltinFunction const> builtin;
+		std::reference_wrapper<yul::FunctionCall const> functionCall;
+		/// Number of proper arguments with a position on the stack, excluding literal arguments.
+		size_t arguments = 0;
+	};
+	struct FunctionCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<Scope::Function const> function;
+		std::reference_wrapper<yul::FunctionCall const> functionCall;
+	};
+	struct Assignment
+	{
+		std::shared_ptr<DebugData const> debugData;
+		/// The variables being assigned to also occur as ``output`` in the ``Operation`` containing
+		/// the assignment, but are also stored here for convenience.
+		std::vector<VariableSlot> variables;
+	};
+
+	struct Operation
+	{
+		/// Stack slots this operation expects at the top of the stack and consumes.
+		Stack input;
+		/// Stack slots this operation leaves on the stack as output.
+		Stack output;
+		std::variant<FunctionCall, BuiltinCall, Assignment> operation;
+	};
+
+	struct FunctionInfo;
+	/// A basic control flow block containing ``Operation``s acting on the stack.
+	/// Maintains a list of entry blocks and a typed exit.
+	struct BasicBlock
+	{
+		std::vector<BasicBlock*> entries;
+		std::vector<Operation> operations;
+		struct MainExit {};
+		struct ConditionalJump
+		{
+			StackSlot condition;
+			BasicBlock* nonZero = nullptr;
+			BasicBlock* zero = nullptr;
+		};
+		struct Jump
+		{
+			BasicBlock* target = nullptr;
+			/// The only backwards jumps are jumps from loop post to loop condition.
+			bool backwards = false;
+		};
+		struct FunctionReturn { CFG::FunctionInfo* info = nullptr; };
+		struct Terminated {};
+		std::variant<MainExit, Jump, ConditionalJump, FunctionReturn, Terminated> exit = MainExit{};
+	};
+
+	struct FunctionInfo
+	{
+		std::shared_ptr<DebugData const> debugData;
+		Scope::Function const& function;
+		BasicBlock* entry = nullptr;
+		std::vector<VariableSlot> parameters;
+		std::vector<VariableSlot> returnVariables;
+	};
+
+	/// The main entry point, i.e. the start of the outermost Yul block.
+	BasicBlock* entry = nullptr;
+	/// Subgraphs for functions.
+	std::map<Scope::Function const*, FunctionInfo> functionInfo;
+	/// List of functions in order of declaration.
+	std::list<Scope::Function const*> functions;
+
+	/// Container for blocks for explicit ownership.
+	std::list<BasicBlock> blocks;
+	/// Container for creates variables for explicit ownership.
+	std::list<Scope::Variable> ghostVariables;
+	/// Container for creates calls for explicit ownership.
+	std::list<yul::FunctionCall> ghostCalls;
```

Comment:
What are ghost calls and variables?

Reply:
I didn't document that anywhere :-)? Doesn't look like I did...
I need ghost calls and ghost variables to transform switches to simpler control flow.
Basically
```
switch value
case 0 { ... }
case 1 { ... }
case 2 { ... }
default { ... }
```
is transformed to
```
let ghostVar := value
if eq(ghostVar, 0) // the "eq" calls being the ghost call
{
  ...
}
else if eq(ghostVar, 1) // well, I can't really write "else" in Yul, but conceptually that's what ends up in the CFG
{
 ....
}
else if eq(ghostVar, 2)
{
 ....
}
else
{
  ...
}
```
So here I need to generate one variable and three calls to ``eq`` for that.
The variable itself is particularly important if the switch value is a function call (which may have side-effects, so I may not be able to just duplicate it).